File: backend/src/main/java/com/interviewer/service/AuthService.java
package com.interviewer.service;

import com.interviewer.config.JwtUtil;
import com.interviewer.dto.AuthResponse;
import com.interviewer.dto.LoginRequest;
import com.interviewer.dto.RegisterRequest;
import com.interviewer.model.User;
import com.interviewer.repository.UserRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;

@Service
public class AuthService {

    @Autowired
    private UserRepository userRepository;

    @Autowired
    private PasswordEncoder passwordEncoder;

    @Autowired
    private JwtUtil jwtUtil;

    /**
     * Register a new user
     */
    public AuthResponse register(RegisterRequest request) {
        // Check if email already exists
        if (userRepository.existsByEmail(request.getEmail())) {
            throw new RuntimeException("Email already registered");
        }
        if (userRepository.existsByUsername(request.getUsername())) {
            throw new RuntimeException("Username already taken");
        }

        // Create user with hashed password
        User user = new User();
        user.setUsername(request.getUsername());
        user.setEmail(request.getEmail());
        user.setPasswordHash(passwordEncoder.encode(request.getPassword()));
        user.setFullName(request.getFullName());

        User savedUser = userRepository.save(user);

        // Generate JWT
        String token = jwtUtil.generateToken(savedUser.getId(), savedUser.getUsername());

        AuthResponse response = new AuthResponse();
        response.setToken(token);
        response.setUsername(savedUser.getUsername());
        response.setFullName(savedUser.getFullName());
        response.setUserId(savedUser.getId());
        response.setMessage("Registration successful");
        return response;
    }

    /**
     * Login user
     */
    public AuthResponse login(LoginRequest request) {
        User user = userRepository.findByEmail(request.getEmail())
                .orElseThrow(() -> new RuntimeException("Invalid email or password"));

        if (!passwordEncoder.matches(request.getPassword(), user.getPasswordHash())) {
            throw new RuntimeException("Invalid email or password");
        }

        String token = jwtUtil.generateToken(user.getId(), user.getUsername());

        AuthResponse response = new AuthResponse();
        response.setToken(token);
        response.setUsername(user.getUsername());
        response.setFullName(user.getFullName());
        response.setUserId(user.getId());
        response.setMessage("Login successful");
        return response;
    }
}

File: backend/src/services/authService.js
/**
 * authService.js — Authentication service
 * Handles user registration and login with bcrypt + JWT.
 * Direct port of Java AuthService.
 */

const bcrypt = require('bcryptjs');
const User = require('../models/User');
const { generateToken } = require('../middleware/auth');

/**
 * Register a new user
 */
async function register(username, email, password, fullName) {
  // Check if email already exists
  const existingEmail = await User.findOne({ email });
  if (existingEmail) {
    throw new Error('Email already registered');
  }

  // Check if username already exists
  const existingUsername = await User.findOne({ username });
  if (existingUsername) {
    throw new Error('Username already taken');
  }

  // Create user with hashed password
  const passwordHash = await bcrypt.hash(password, 10);

  const user = new User({
    username,
    email,
    passwordHash,
    fullName
  });

  const savedUser = await user.save();

  // Generate JWT
  const token = generateToken(savedUser._id.toString(), savedUser.username);

  return {
    token,
    username: savedUser.username,
    fullName: savedUser.fullName,
    userId: savedUser._id.toString(),
    message: 'Registration successful'
  };
}

/**
 * Login user
 */
async function login(email, password) {
  const user = await User.findOne({ email });
  if (!user) {
    throw new Error('Invalid email or password');
  }

  const isMatch = await bcrypt.compare(password, user.passwordHash);
  if (!isMatch) {
    throw new Error('Invalid email or password');
  }

  const token = generateToken(user._id.toString(), user.username);

  return {
    token,
    username: user.username,
    fullName: user.fullName,
    userId: user._id.toString(),
    message: 'Login successful'
  };
}

module.exports = { register, login };
